src/loader.c

```diff
@@ -95,12 +96,96 @@ static char *skip_bom(char *string)
 	return string;
 }
 
+/* we have made changes so that SUPER long lines (greater than LINE_BUFFER_SIZE
+ * will now get 'trimmed' when put into the .pot file. Here is the trimming
+ * method:
+ *    input:    $hashtype$abcdefghijk..........qrstuvwxzy$something$else
+ *    pot:      $hashtype$abcdefghijk.......$SOURCE_HASH$<md5 of full hash>$SOURCE_BIN$<binary() of hash in hex>
+ * this way we can fully compare this .pot record (against the full input line)
+ */
+static int ldr_pot_source_cmp(const char *pot_entry, const char *full_source) {
+	MD5_CTX ctx;
+	unsigned char srcH[16], potH[16];
+	const char *p;
+	if (!strcmp(pot_entry, full_source))
+		return 0;
+	p = strstr(pot_entry, "$SOURCE_HASH$");
+	if (!p)
+		return 1; /* can not be a match */
+	if (strncmp(full_source, pot_entry, p-pot_entry))
+		return 1; /* simple str compare shows they are not the same */
+	/* ok, this could be a match.  Now we check the hashes */
+	MD5_Init(&ctx);
+	MD5_Update(&ctx, full_source, strlen(full_source));
+	MD5_Final(srcH, &ctx);
+	p += 13;
+	base64_convert(p, e_b64_hex, 32, potH, e_b64_raw, 16, 0);
+	if (!memcmp(srcH, potH, 16))
+		return 0;
+	return 1;
+}
+
+/*
+ * not static function.  Used by cracker.c This function builds a proper
+ * source line to be written to the .pot file. This string MAY be the
+ * original source line, OR it may be a chopped down (shortened) source
+ * line with some extra data tacked on. However, it will always be shorter
+ * or equal to LINE_BUFFER_SIZE
+ */
+#define BIN_LEN (sizeof(LDR_TRIMMED_POT_BIN_SIG)+1)
+const char *ldr_pot_source(const char *full_source,
+                           char buffer[LINE_BUFFER_SIZE+1],
+                           void *bin, int blen)
+{
+	MD5_CTX ctx;
+	int len;
+	char *p = buffer;
+	unsigned char mbuf[16];
+
+	if (strlen(full_source) <= LINE_BUFFER_SIZE)
+		return full_source;
+
+	/* we create a .pot record that is LINE_BUFFER_SIZE lone
+	 * but that has a hash of the full source, and the binary
+	 * from the full source appened. Both of those strings are
+	 * later needed for .pot comparison, and .pot removal logic
+	 */
+
+	/* 13=len sig1, 32=len hash */
+	len = LINE_BUFFER_SIZE - 13 - 32 - BIN_LEN - blen*2;
+	memcpy(p, full_source, len);
+	p += len;
+	memcpy(p, "$SOURCE_HASH$", 13);
+	p += 13;
+	MD5_Init(&ctx);
+	MD5_Update(&ctx, full_source, strlen(full_source));
+	MD5_Final(mbuf, &ctx);
+	base64_convert(mbuf, e_b64_raw, 16, p, e_b64_hex, 33, 0);
+	p += 32;
+	memcpy(p, LDR_TRIMMED_POT_BIN_SIG, BIN_LEN);
+	p += BIN_LEN;
+	base64_convert(bin, e_b64_raw, blen, p,
+	               e_b64_hex, blen*2 + 1, 0);
+	return buffer;
+}
+
+/* returns true or false depending if this ciphertext is a trimmed .pot line */
+int ldr_isa_pot_source(const char *ciphertext) {
+	if (strlen(ciphertext) <= LINE_BUFFER_SIZE) {
+		const char *cp = strstr(ciphertext, "$SOURCE_HASH$");
+		if (!cp)
+			return 0;
+		return !!strstr(cp, "$SOURCE_BIN$");
+	}
+	return 0;
+}
+
```

Comment:
The way this function is implemented above paints us into a corner if we ever want to change LINE_BUFFER_SIZE. So despite the performance impact, we might want to always do the `strstr(ciphertext, "$SOURCE_HASH$")`.

But this function won't ever be called unless we're actually using a format with potentially long lines, right? I certainly hope it will not.